### PR TITLE
BUG: int overflow in reshape, fixes #7455, fixes #7293

### LINKED ISSUE #7293: Clarify the error message on reshape failure
Currently, if incompatible values are passed to `np{,.ndarray}.reshape`, the error message is a rather unhelpful "total size of new array must be unchanged".  It would be nicer if this was instead something like "Cannot reshape array of size N into shape (X, Y, Z)".
(Because `reshape` is written in C, if something like `func().reshape(...)` fails, then you can't even check for the size of `func()` via `pdb` and you actually need to edit your code there.)

### LINKED ISSUE #7455: Segfault in array reshaping due to overflow
The following should segfault on any 64 bit NumPy:

```Python
a = np.ones(20)[::2]
a.reshape(2, 13, 419, 691, 823, 2977518503)
```

The trick is that, due to overflow:

```Python
>>> (np.int64(2) * np.int64(13) * np.int64(419) * np.int64(691) * np.int64(823) *
...  np.int64(2977518503))
10
```

so the sizes of the arrays actually "match."

The values have been found by factoring `2**63 + 5`, then adding the leading 2 factor to make the whole result positive after overflow. For 32 bit systems, this should do the trick (untested):

```Python
a = np.ones(20)[::2]
a.reshape(2, 7, 7, 43826197)
```

The slicing of `a` is needed to have a non-contiguous array, so that the code of `PyArray_NewShape` calls [`_attempt_nocopy_reshape`](https://github.com/numpy/numpy/blob/maintenance/1.11.x/numpy/core/src/multiarray/shape.c#L358). Trying the same with `np.ones(10)` gives a proper "array too big" error.

There are computations of array size without proper overflow checking both in `_attempt_nocopy_reshape` and `_fix_unknown_dimension`. It may be enough to fix the check in `_fix_unknown_dimension`, as an error there will prevent `_attempt_nocopy_reshape` from being reachable, although it would be good to then remove the redundant size match check and document that it depends on the sizes actually matching. A test with the above examples should keep us safe from ignoring our own comments in the code.

And that's as far as I have gone into debugging this, will probably try fixing it over Easter if no one beats me to it.

### PR DESCRIPTION
**Pull request by jaimefrio**

### CODE REVIEW

File: numpy/core/src/multiarray/shape.c [hunk 1]
```diff
@@ -508,26 +488,26 @@ _fix_unknown_dimension(PyArray_Dims *newshape, npy_intp s_original)
...
                 PyErr_SetString(PyExc_ValueError,
-                                "can only specify one"  \
-                                " unknown dimension");
+                                "can only specify one unknown dimension");
                 return -1;
             }
         }
-        else {
-            s_known *= dimensions[i];
+        else if(npy_mul_with_overflow_intp(&s_known, s_known, dimensions[i])) {
```
Comment (seberg) [inline numpy/core/src/multiarray/shape.c]: Missing space, but everything looks good too me.

**Comment by jaimefrio** [comment]
Fixed the spacing, thanks for the review. From looking at all those shape functions, I have the feeling that there has to be a simpler, cleaner way of rewriting all of that. Maybe some other day...

**Comment by seberg** [comment]
Hehe, probably, though they also all have their quirks you have to keep alive probably ;).

**Comment by seberg** [comment]
Anyway, thanks!

**Pull request merged**